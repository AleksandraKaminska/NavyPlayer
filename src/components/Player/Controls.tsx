import React, { useState, useEffect, useContext } from 'react'
import { ReactSVG } from 'react-svg'
import { Button, Space, Tooltip } from 'antd'
import { DispatchContext, StateContext } from '../../context/Context'
import { searchArtistInfo } from '../../helpers/search'
import { random } from '../../helperFunctions'
import PlayIcon from './play.svg'
import RewindIcon from './rewind.svg'
import ForwardIcon from './forward.svg'
import { StateType } from '../../reducers'
const { DZ } = window

function Controls({ repeat }: { repeat: boolean }) {
  const dispatch = useContext<React.Dispatch<any>>(DispatchContext)
  const state = useContext<StateType>(StateContext)
  const [isPlaying, setIsPlaying] = useState<boolean>(false)

  const changeIsPlaying = (): void => {
    setIsPlaying(!isPlaying)
    isPlaying ? DZ?.player.pause() : DZ?.player.play()
  }

  const changeTrack = (): void => {
    setIsPlaying(true)
    DZ?.player.pause()
    repeat ? DZ?.player.playTracks([state.track?.id]) : random(state, dispatch)
  }

  const rewind = (): void => {
    setIsPlaying(true)
    DZ?.player.pause()
    if (state.previousTracks?.length && state.track) {
      const previousTrack = state.previousTracks[state.previousTracks.length - 1]
      dispatch({ type: 'CHANGE_TRACK', payload: previousTrack })
      dispatch({ type: 'PREV_TRACK', payload: previousTrack })
      searchArtistInfo(previousTrack, dispatch)
    }
  }

  useEffect(() => {
    DZ?.ready(() => {
      changeTrack()
    })
  }, [])

  DZ?.Event.subscribe('track_end', () => {
    changeTrack()
  })

  return (
    <Space className="Controls" align="center">
      <Tooltip title="Previous">
        <Button onClick={rewind} type="text" disabled={state.previousTracks?.length === 0}>
          <ReactSVG src={RewindIcon} />
        </Button>
      </Tooltip>
      <Tooltip title={isPlaying ? 'Pause' : 'Play'}>
        <Button onClick={changeIsPlaying} type="text">
          {isPlaying ? (
            <img src="/assets/images/pause.png" alt="pause button" width="24px" style={{ verticalAlign: 'baseline' }} />
          ) : (
            <ReactSVG src={PlayIcon} />
          )}
        </Button>
      </Tooltip>
      <Tooltip title="Next">
        <Button onClick={changeTrack} type="text">
          <ReactSVG src={ForwardIcon} />
        </Button>
      </Tooltip>
    </Space>
  )
}

export default Controls
