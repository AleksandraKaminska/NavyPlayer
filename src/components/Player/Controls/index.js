import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as actions from 'actions';
import store from 'store';
import { searchArtistInfo, random } from 'helperFunctions';
import Rewind from './rewind';
import Forward from './forward';
import icons from 'const';

const { DZ } = window;

const promise = new Promise((resolve, reject) => {
    true ? resolve("Stuff worked!") : reject(Error("It broke"));
});

class Controls extends Component {
    constructor() {
        super()
        this.state = {
            isPlaying: false,
            repeat: false,
            mouseDown: false
        }
        this.progress = React.createRef()
        this.line = React.createRef()
        this.ball = React.createRef()
    }

    setVolume = event => {
        const vol = this.getVolPercent(event)
        this.line.current.style.height = this.ball.current.style.bottom = vol + '%'
        DZ.player.setVolume(vol)
    }

    onMouseUp(event) {
        if (this.state.mouseDown) {
            this.setState({ mouseDown: false })
            this.setVolume(event)
        }
    }
    
    onMouseDown(event) {
        this.setState({ mouseDown: true })
        this.setVolume(event)
    }

    onMouseMove(event) {
        if (!this.state.mouseDown) {
          return;
        }
        this.setVolume(event)
    }
    
    onMouseLeave() {
        this.setState({ mouseDown: false })
    }

    getVolPercent({ currentTarget, clientY }) {
        const { top, bottom, height } = currentTarget.getBoundingClientRect();
        if (clientY <= top) {
            return 100;
        }
        if (clientY >= bottom) {
            return 0;
        }
        const vol = 100 - (clientY - top) / height * 100;
        if (vol <= 5.5) {
            return 0;
        }  
        return vol;
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.track.id && nextProps.track !== this.props.track ) {
            this.setState({ isPlaying: true });
        }
    }

    changeIsPlaying = () => {
        this.setState(
            { isPlaying: !DZ.player.isPlaying() },
            () => DZ.player.isPlaying() ? DZ.player.pause() : DZ.player.play()
        );
    }

    changeTrack = () => {
        this.setState({
            isPlaying: true
        }, () => {
            DZ.player.pause()
            this.state.repeat ? DZ.player.playTracks([this.props.track.id]) : random(this.props)
        });
    }

    rewind = () => {
        this.setState({
            isPlaying: true
        }, () => {
            DZ.player.pause();
            const { prev } = this.props;
            if (prev.length) {
                promise.then(result => {
                    store.dispatch(actions.changeTrackAction(prev.slice(-1)[0]));
                    store.dispatch(actions.prevTrackAction(prev.slice(-1)[0]));
                    DZ.player.playTracks([this.props.track.id]);
                    searchArtistInfo(this.props.track);
                }, err => console.log(err));
            } else {
                random(this.props)
            }
        });
    }

    render() {
        return (
            <div className="controls">
                <button>
                    <img
                        src={icons.volume}
                        alt=""
                    />
                    <div 
                        className="volumeSlider"
                        ref={this.progress}
                        onMouseUp={ev => this.onMouseUp(ev)}
                        onMouseDown={ev => this.onMouseDown(ev)}
                        onMouseLeave={ev => this.onMouseLeave(ev)} 
                        onMouseMove={ev => this.onMouseMove(ev)}
                    >
                        <div className="volumeSlider__lineWrapper">
                            <div className="volumeSlider__bgLine" />
                            <div className="volumeSlider__currentLine" ref={this.line} />
                            <div className="volumeSlider__ball" ref={this.ball} />
                        </div>
                    </div>
                </button>
                <button>
                    <Rewind onClick={this.rewind} />
                </button>
                <button onClick={this.changeIsPlaying}>
                    <img src={ this.state.isPlaying ? '/assets/images/pause.png' : icons.play } alt="" />
                </button>
                <button>
                    <Forward onClick={this.changeTrack} />
                </button>
                <button>
                    <img
                        src={ this.state.repeat ? icons.repeatBlue : icons.repeatWhite }
                        onClick={() => {
                            DZ.player.setRepeat(this.state.repeat ? 0 : 2)
                            this.setState({ repeat: !this.state.repeat })
                        }}
                        alt=""
                    />
                </button>
            </div>
        );
    }
}

const mapStateToProps = ({ track, prev, chosenPlaylist, album, flow, artist, artistPlaylist }) => ({
    track,
    prev,
    chosenPlaylist,
    album,
    flow,
    artist,
    artistPlaylist
});

export default connect(mapStateToProps)(Controls);
