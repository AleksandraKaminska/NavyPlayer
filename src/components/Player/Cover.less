@import '~antd/dist/antd.dark.less'; // Introduce the official dark less style entry file

.Cover {
  color: @white;
  text-align: left;
  .title {
    font-weight: 900;
  }
}