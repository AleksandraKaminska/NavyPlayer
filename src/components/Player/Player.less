.Player {
  .align-start {
    display: flex;
    justify-content: flex-start;
  }
  .button {
    color: white;
    &:disabled {
      color: grey;
    }
  }
}