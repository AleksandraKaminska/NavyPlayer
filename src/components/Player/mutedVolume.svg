<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" id="Capa_1" x="0px" y="0px" viewBox="0 0 79.144 79.144" style="enable-background:new 0 0 79.144 79.144;" xml:space="preserve" width="1.4em" height="1.4em">
	<g>
		<path d="M0,28.625v22.062c0,3.744,3.257,6.41,6.646,6.41h9.749c2.053,1,11.604,8.503,19.806,15.51    c0.369,0.315,0.831,0.574,1.299,0.574c0.284,0,0.319-0.013,0.587-0.136C38.796,72.718,39,72.057,39,71.276V8.035    c0-0.788-0.212-1.502-0.931-1.824c-0.718-0.323-1.436-0.385-2.023,0.14c-8.179,7.271-17.586,15.746-19.665,15.746H6.647    C3.085,22.097,0,25.063,0,28.625z M4,28.625c0-1.324,1.322-2.528,2.646-2.528h9.928c0.742,0,2.426,0.381,18.426-13.631v54.493    C19,53.478,17.312,53.097,16.574,53.097H6.646C5.362,53.097,4,52.035,4,50.687V28.625z" fill="#FFFFFF"/>
	</g>
</svg>