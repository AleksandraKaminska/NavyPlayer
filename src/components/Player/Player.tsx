import React, { useState } from 'react'
import Controls from './Controls'
import Progress from './Progress'
import Volume from './Volume'
import SmallCover from './SmallCover'
// import './style.scss'

const Player: React.FC = () => {
  const [repeat, setRepeat] = useState<boolean>(false)

  const toggleRepeat = () => setRepeat(!repeat)

  return (
    <section className="player">
      <SmallCover />
      <Controls repeat={repeat} />
      <Progress />
      <Volume repeat={repeat} changeRepeat={toggleRepeat} />
    </section>
  )
}

export default Player
