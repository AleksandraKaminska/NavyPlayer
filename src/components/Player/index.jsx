import React, { Component } from 'react'
import Controls from './Controls'
import Progress from './Progress'
import Volume from './Volume'
import SmallCover from './SmallCover'
import './style.scss'

class Player extends Component {
  constructor() {
      super()
      this.state = {
          repeat: false
      }
  }

  changeRepeat = () => {
    this.setState({ repeat: !this.state.repeat })
  }

  render() {
    return (
      <section className='player'>
        <SmallCover />
        <Controls repeat={this.state.repeat} />
        <Progress />
        <Volume repeat={this.state.repeat} changeRepeat={this.changeRepeat} />
      </section>
    )
  }
}

export default Player
