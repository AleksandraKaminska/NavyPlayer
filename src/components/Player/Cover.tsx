import React, { useContext } from 'react'
import { Link } from 'react-router-dom'
import { Space } from 'antd'
import { Context } from '../../context/Context'
import './Cover.less'

const Cover = () => {
  const {
    state: { track }
  } = useContext(Context)

  return (
    <Space className="Cover">
      <img src={track?.album?.cover_small} alt={track?.album?.title} />
      <div className="data">
        <Link to={`/album/${track?.album?.id}`} className="title">
          {track?.title_short}
        </Link>
        <Link to={`/artist/${track?.artist.id}`} className="name">
          {track?.artist.name}
        </Link>
      </div>
    </Space>
  )
}

export default Cover
