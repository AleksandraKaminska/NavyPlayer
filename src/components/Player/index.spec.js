import React from 'react'
import Player from './index'
import renderer from 'react-test-renderer'

jest.mock('./Controls', () => 'Controls')
jest.mock('./Progress', () => 'Progress')
jest.mock('./Volume', () => 'Volume')
jest.mock('./SmallCover', () => 'SmallCover')

describe('Player', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Player />)
    expect(tree).toMatchSnapshot()
  })
})
