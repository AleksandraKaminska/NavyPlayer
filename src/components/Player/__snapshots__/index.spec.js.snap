// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<section
  className="player"
>
  <SmallCover />
  <Controls
    repeat={false}
  />
  <Progress />
  <Volume
    changeRepeat={[Function]}
    repeat={false}
  />
</section>
`;
